cub/test/catch2_test_device_adjacent_difference_substract_left.cu

```diff
@@ -0,0 +1,302 @@
+{
+  using type = typename c2h::get<0, TestType>;
+
+  const int num_items = GENERATE_COPY(take(2, random(1, 1000000)));
+  thrust::device_vector<type> in(num_items);
+  thrust::device_vector<type> out(num_items);
+  c2h::gen(CUB_SEED(2), in);
+
+  thrust::host_vector<type> h_in = in;
+  thrust::host_vector<type> reference(num_items);
+  std::adjacent_difference(h_in.begin(), h_in.end(), reference.begin(), cust_diff{});
+
+  adjacent_difference_subtract_left_copy(in.begin(),
+                                         out.begin(),
+                                         num_items,
+                                         cust_diff{});
+
+  REQUIRE(reference == out);
+}
+
+template<class T>
+struct convertible_from_T {
+  T val_;
+
+  convertible_from_T() = default;
+  __host__ __device__ convertible_from_T(const T& val) noexcept : val_(val) {}
+  __host__ __device__ convertible_from_T& operator=(const T& val) noexcept {
+    val_ = val;
+  }
+  // Converting back to T helps satisfy all the machinery that T supports
+  __host__ __device__ operator T() const noexcept { return val_; }
+};
+
+CUB_TEST("DeviceAdjacentDifference::SubtractLeftCopy works with a different output type", "[device][adjacent_difference]", types)
+{
+  using type = typename c2h::get<0, TestType>;
+
+  const int num_items = GENERATE_COPY(take(2, random(1, 1000000)));
+  thrust::device_vector<type> in(num_items);
+  thrust::device_vector<convertible_from_T<type>> out(num_items);
+  c2h::gen(CUB_SEED(2), in);
+
+  thrust::host_vector<type> h_in = in;
+  thrust::host_vector<type> reference(num_items);
+  std::adjacent_difference(h_in.begin(), h_in.end(), reference.begin(), cust_diff{});
+
+  adjacent_difference_subtract_left_copy(in.begin(),
+                                         out.begin(),
+                                         num_items,
+                                         cust_diff{});
+
+  REQUIRE(reference == out);
+}
+
+struct check_difference {
+  template<class T>
+  __host__ __device__ T operator()(const T& lhs, const T& rhs) const noexcept {
+    const T result = lhs - rhs;
+    assert(result == 1);
+    return result;
+  }
+};
+
+CUB_TEST("DeviceAdjacentDifference::SubtractLeftCopy works with large indexes", "[device][adjacent_difference]")
+{
+  const cuda::std::size_t num_items = 1ll << 33;
+  adjacent_difference_subtract_left_copy(thrust::counting_iterator<cuda::std::size_t>{0},
+                                         thrust::discard_iterator<>{},
+                                         num_items,
+                                         check_difference{});
+}
+
+struct invocation_counter {
+  cuda::std::atomic_ref<cuda::std::size_t> counts_;
+
+  __host__ __device__ explicit invocation_counter(cuda::std::size_t* counts) : counts_(*counts) {}
+
+  template<class T>
+  __host__ __device__ T operator()(const T& lhs, const T& rhs) const noexcept {
```

Comment:
important: I'd like to avoid accidental invocation of the function object on the host:
```suggestion
  __device__ T operator()(const T& lhs, const T& rhs) const noexcept {
```